runtime/src/accounts.rs

```diff
@@ -314,36 +346,68 @@ impl Accounts {
     /// Slow because lock is held for 1 operation instead of many
     pub fn store_slow(&self, fork: Fork, pubkey: &Pubkey, account: &Account) {
         let mut accounts = HashMap::new();
-        accounts.insert(pubkey, (account, 0));
+        accounts.insert(pubkey, (account, None));
         self.accounts_db.store(fork, &accounts);
     }
 
     fn lock_account(
         locks: &mut HashSet<Pubkey>,
-        keys: &[&Pubkey],
+        credit_only_locks: &mut HashMap<Pubkey, Arc<AtomicU64>>,
+        message: &Message,
         error_counters: &mut ErrorCounters,
-    ) -> Result<()> {
-        // Copy all the accounts
-        for k in keys {
+    ) -> Result<CreditOnlyLocks> {
+        let (credit_debit_keys, credit_only_keys) = message.get_account_keys_by_lock_type();
+
+        for k in credit_debit_keys.iter() {
+            if locks.contains(k) || credit_only_locks.contains_key(k) {
+                error_counters.account_in_use += 1;
+                debug!("Account in use: {:?}", k);
+                return Err(TransactionError::AccountInUse);
+            }
+        }
+        for k in credit_only_keys.iter() {
             if locks.contains(k) {
                 error_counters.account_in_use += 1;
                 debug!("Account in use: {:?}", k);
                 return Err(TransactionError::AccountInUse);
             }
         }
-        for k in keys {
-            locks.insert(**k);
+
+        for k in credit_debit_keys {
+            locks.insert(*k);
+        }
+        let mut lock_cache: CreditOnlyLocks = vec![];
+        for k in credit_only_keys {
+            if !credit_only_locks.contains_key(k) {
+                credit_only_locks.insert(*k, Arc::new(AtomicU64::new(0)));
+            }
+            let lock = credit_only_locks.get(k).unwrap().clone();
```

Comment:
This block can be simplified to `let lock = credit_only_locks.entry(k).or_insert(Arc::new(AtomicU64::new(0)));`